ftdc/ftdc.go

```diff
@@ -439,6 +459,107 @@ func (ftdc *FTDC) getWriter() (io.Writer, error) {
 	return ftdc.outputWriter, nil
 }
 
+func (ftdc *FTDC) fileDeleter() {
+	for {
+		select {
+		// The fileDeleter's goroutine lifetime should match the robot/FTDC lifetime. Borrow the
+		// `readStatsWorker`s context to track that.
+		case <-ftdc.readStatsWorker.Context().Done():
+			return
+		case <-time.After(time.Second):
+		}
+
+		if err := ftdc.checkAndDeleteOldFiles(); err != nil {
+			ftdc.logger.Warnw("Error checking FTDC files", "err", err)
+		}
+	}
+}
+
+// fileTime pairs a file with a time value.
+type fileTime struct {
+	name string
+	time time.Time
+}
+
+func (ftdc *FTDC) checkAndDeleteOldFiles() error {
+	var files []fileTime
+
+	// Walk the `ftdcDir` and gather all of the found files into the captured `files` variable.
+	err := filepath.Walk(ftdc.ftdcDir, filepath.WalkFunc(func(path string, info fs.FileInfo, walkErr error) error {
+		if !strings.HasSuffix(path, ".ftdc") {
+			return nil
+		}
+
+		if walkErr != nil {
+			ftdc.logger.Warnw("Unexpected walk error. Continuing under the assumption any actual* problem will",
+				"be caught by the assertions.", "err", walkErr)
+			return nil
+		}
+
+		parsedTime, err := parseTimeFromFilename(path)
+		if err == nil {
+			files = append(files, fileTime{path, parsedTime})
+		} else {
+			ftdc.logger.Warnw("Error parsing time from FTDC file", "filename", path)
+		}
+		return nil
+	}))
+	if err != nil {
+		return err
+	}
+
+	if len(files) <= ftdc.maxNumFiles {
+		// We have yet to hit our file limit. Keep all of the files.
+		ftdc.logger.Debugw("Inside the budget for ftdc files", "numFiles", len(files), "maxNumFiles", ftdc.maxNumFiles)
+		return nil
+	}
+
+	slices.SortFunc(files, func(left, right fileTime) int {
+		// Sort in descening order. Such that files indexed first are safe. This eases walking the
+		// slice of files.
+		return right.time.Compare(left.time)
+	})
+
+	// If we, for example, have 30 files and we want to keep the newest 10, we delete the trailing
+	// 20 files.
+	for _, file := range files[ftdc.maxNumFiles:] {
+		ftdc.logger.Debugw("Deleting aged out FTDC file", "filename", file.name)
+		if err := os.Remove(file.name); err != nil {
+			ftdc.logger.Warnw("Error removing FTDC file", "filename", file.name)
+		}
+	}
+
+	return nil
+}
+
+// filenameTimeRe matches the files produced by ftdc. Filename <-> regex parity is exercised by file
+// deletion testing. Filename generation uses padding such that we can rely on there before 2/4
+// digits for every numeric value.
```

Comment:
Could you put an example of what this regex matches in this comment?

Reply:
done